rfcs/0000-python-sdk.md

```diff
@@ -0,0 +1,276 @@
+
+### Create an importer from a csv
+Importers allow for the creation of jobs to ingest feature data into Feast. Importers are generally (but not necessarily) created from existing data sets (BQ, CSV, Pandas dataframe) in order to discover feature names.
+
+```python
+driver_importer = importer.from_csv('driver_features.csv', granularity=granularity.DAY, entity='driver', owner='owner@email.com')
+```
+
+### [Alternative] Create an importer from a BigQuery table
+
+```python
+driver_importer_from_bq = importer.from_bq('project.dataset.drivers_table', granularity=granularity.DAY, entity='driver', owner='owner@email.com')
+```
+
+### [Alternative] Create an importer from a Pandas dataframe
+
+```python
+my_pandas_df = pd.read_csv('driver_features.csv')
+driver_importer_from_df = importer.from_df(my_pandas_df, granularity=granularity.DAY, entity='driver', owner='owner@email.com')
+```
+
+### Preview the dataframe loaded by the importer
+The `head` method prints out the first 5 rows of the Pandas dataframe.
+
+```python
+driver_importer.df.head()
+```
+
+### Change the local file path in the importer to GCS
+In order to ingest data from a bucket, it is necessary to modify the importer's path configuration to the location of the file that needs to be imported.
+
+```python
+driver_importer.options.path = 'gs://my-bucket-location/driver_features.csv'
+```
+
+### Describe the importer
+
+```python
+driver_importer.describe()
+```
+
+```yml
+type: file
+options:
+  format: csv
+  path: gs://my-bucket-location/driver_features.csv
+entities:
+  - driver
+schema:
+  entityIdColumn: driver_id
+  timestampColumn: ts
+  fields:
+    - name: timestamp
+    - name: customer_id
+    - name: completed
+      featureId: completed
+    - name: avg_distance_completed
+      featureId: avg_distance_completed
+    - name: avg_customer_distance_completed
+      featureId: avg_customer_distance_completed
+    - name: avg_distance_cancelled
+      featureId: avg_distance_cancelled
+```
+
+### Submit the import job
+The import job loads the CSV from GCS into Feast. It automatically registers entities and features with Feast during submission.
+
+```python
+fs.apply(driver_importer, create_entity=True, create_features=True)
+```
+_starting import..._  
+_10%_   
+_50%_   
+_100%_    
+_10 rows imported successfully_ 
+
+### Write out specification files for later use
+
+```python
+driver_importer.dump("driver_feature_import.yaml")
+customer_entity.dump("customer_entity.yaml")
+customer_age.dump("customer_entity.yaml")
+```
+
+### Create a “feature set” which can be used to query both training data and serving data.
+The feature set is simply an object that locally tracks which entity, granularity, and features you are interested in.
+
+```python
+feature_set = fs.create_feature_set(entity='driver', granularity='minute', features=['latitude', 'longitude', 'event_time'])
```

Comment:
Do you think it make more sense to have FeatureSet creation outside of client, and then to get training/serving data we pass the instance into the corresponding client method?

```
# create feature set
feature_set = FeatureSet(entity='driver', granularity='minute', features=['latitude', 'longitude', 'event_time'])

# training dataset
dataset_info = fs.create_training_dataset(feature_set, start_date='2018-01-01', end_date='2018-02-01')
dataset_info.download(dest=file_path, type='feather')
df = dataset_info.download_to_df()

# serving
feature_data = fs.get_serving_data(feature_set, keys, type='last')
```